keps/sig-api-machinery/20190425-structural-openapi.md

```diff
@@ -0,0 +1,812 @@
+* OpenAPI v3 is a very open and wide spec that can define crazy things using logical operators (`not(anyOf(oneof ...))`), leading to complicated semantics, making advanced CRD features difficult or even impossible to implement.
+
+### Goals
+
+* Restrict CRD validation OpenAPI v3 schemas to give one canonical way to express objects we intend to support as Kubernetes-like types.
+* Simplify the semantics of CRD validation OpenAPI v3 schemas to make implementation of advanced feature like pruning, defaulting, server-side apply and read-only validation considerably less difficult.
+* Support embedding of existing API types into a CRD via RawExtensions or via Golang type reuse (e.g. of PodSpecs).
+* Allow a rich OpenAPI v3 validation schema for CustomResource validation, but allow us to programmatically derive the restricted spec for all other purposes.
+* Prospective goal: have a canonical mapping from a restricted schema to Golang types and Protobuf v3.
+
+### Non-Goals
+
+* a human readable spec and convenience to write. We focus on machine readability and clear and simple semantics.
+* model all kind of crazy JSON objects. We focus on Kubernetes-like objects with an OpenAPI schema derived via code generation (crd-gen) from Golang types.
+
+### Context
+
+We assume the following context of OpenAPI schemas in the Kubernetes architecture and ecosystem:
+
+#### Consumers of the restricted OpenAPI Schema
+
+* server side validation (consumes structural schema including value validation)
+* OpenAPI publishing
+  * client side validation
+  * kubectl explain
+  * client generators (https://github.com/kubernetes-client)
+  * doc generator
+  * kube-aggregator
+  * IDEs like IntelliJ for input completion
+* server side apply
+* pruning
+* defaulting
+* possibly in the future:
+  * server side read-only update validation
+  * Protobuf publishing
+
+#### Producers of the restricted OpenAPI Schema
+
+* kube-openapi's **openapi-gen** mechanically producing go-openapi definitions for native types compiled into kube-apiserver
+* **crd-gen/crd-schema-gen** in https://github.com/kubernetes-sigs/controller-tools mechanically producing CRD validation schemas from Golang types with tags
+* a **developer** adding custom OpenAPI schemas for openapi-gen or crd-gen/crd-schema-gen.
+* a **developer** manually writing complete CRD validation schemas.
+
+![Information flow of OpenAPI schemas in Kubernetes](20190425-structural-openapi-context.png)
+
+### Criteria
+
+To allow only a subset of features of OpenAPI v3, we need to have guidelines for what features we want to restrict and what features we need to permit. We would like to select a single canonical way in OpenAPI to represent these Kubernetes-like types:
+
+* atomic types: primitive types, maps, arrays.
+
+  Examples:
+  * string maps like labels and annotations, but also `map[string]int64`
+  * IPv4 vs IPv6 strings: same fields with different formats
+  * embedded PodSpec
```

Comment:
does an embedded pod spec fit in this category? it has discriminated union types within it (like volumesource), as well as single and multi-key arrays (containers, ports, etc)

Reply:
Does it have any type we cannot express at all? (there are certain things we cannot validate completely, but the base types should be structural.